legacy/src/main/java/org/opensearch/sql/legacy/cursor/DefaultCursor.java

```diff
@@ -70,31 +95,85 @@ public class DefaultCursor implements Cursor {
   /** To get next batch of result */
   private String scrollId;
 
+  /** To get Point In Time */
+  private String pitId;
+
+  /** To get next batch of result with search after api */
+  private SearchSourceBuilder searchSourceBuilder;
+
+  /** To get last sort values * */
+  private Object[] sortFields;
+
   /** To reduce the number of rows left by fetchSize */
   @NonNull private Integer fetchSize;
 
   private Integer limit;
 
+  /**
+   * {@link NamedXContentRegistry} from {@link SearchModule} used for construct {@link QueryBuilder}
+   * from DSL query string.
+   */
+  private static final NamedXContentRegistry xContentRegistry =
+      new NamedXContentRegistry(
+          new SearchModule(Settings.EMPTY, Collections.emptyList()).getNamedXContents());
+
   @Override
   public CursorType getType() {
     return type;
   }
 
   @Override
   public String generateCursorId() {
-    if (rowsLeft <= 0 || Strings.isNullOrEmpty(scrollId)) {
+    if (rowsLeft <= 0 || isCursorIdNullOrEmpty()) {
       return null;
     }
     JSONObject json = new JSONObject();
     json.put(FETCH_SIZE, fetchSize);
     json.put(ROWS_LEFT, rowsLeft);
     json.put(INDEX_PATTERN, indexPattern);
-    json.put(SCROLL_ID, scrollId);
     json.put(SCHEMA_COLUMNS, getSchemaAsJson());
     json.put(FIELD_ALIAS_MAP, fieldAliasMap);
+    if (LocalClusterState.state().getSettingValue(SQL_PAGINATION_API_SEARCH_AFTER)) {
+      json.put(PIT_ID, pitId);
+      String sortFieldValue =
+          AccessController.doPrivileged(
+              (PrivilegedAction<String>)
+                  () -> {
+                    try {
+                      return objectMapper.writeValueAsString(sortFields);
+                    } catch (JsonProcessingException e) {
+                      throw new RuntimeException(
+                          "Failed to parse sort fields from JSON string.", e);
+                    }
+                  });
+      json.put(SORT_FIELDS, sortFieldValue);
+      setSearchRequestString(json, searchSourceBuilder);
+    } else {
+      json.put(SCROLL_ID, scrollId);
+    }
     return String.format("%s:%s", type.getId(), encodeCursor(json));
   }
 
+  private void setSearchRequestString(JSONObject cursorJson, SearchSourceBuilder sourceBuilder) {
+    try {
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
```

Comment:
I don't know if we really need output stream here:

try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
    sourceBuilder.toXContent(builder, ToXContent.EMPTY_PARAMS);
    String jsonString = builder.toString();
    searchRequestBase64 = Base64.getEncoder().encodeToString(jsonString.getBytes(StandardCharsets.UTF_8));
}
can you take this after the PR?